Focus Management added to FormModel/FieldModel

cnrudd opened the pull request:
+ This PR adds some basic focus management to Form/FieldModel
+ This PR add a `boundInput` property to FieldModel for imperative access
+ This PR provides standard implementations of focus/blur/select to HoistInputs.

See toolbox PR, for enhanced forms that allow testing of the new Focus Management methods.

Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [x] Added CHANGELOG entry.
- [x] Reviewed for breaking changes: no breaking changes
- [x] Updated doc comments / prop-types, or determined not required.
- [x] Reviewed and tested on Mobile: implemented equivalent changes on mobile HoistInput components.
- [x] Created Toolbox branch: https://github.com/xh/toolbox/compare/focusBlur?expand=1

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

lbwexler commented on desktop/cmp/input/Select.js:
```diff
@@ -613,7 +634,7 @@ const cmp = hoistCmp.factory(
                 filterOption: model.filterOption,
 
-                ref: model.reactSelectRef
+                ref: composeRefs(model.inputRef, props.inputRef, model.reactSelectRef)
```
why does the model need two identical refs? why are we introducing a new prop ref here?  this seems to me to be cruft.

cnrudd replied:
Fixed.

lbwexler commented:
This looks great!

+ So we will want to add documented methods signatures (with empty implementations) for all of these on HoistInputModel itself.

+ Also would like to provide mobile implementations

+ interesting question as to whether the imperative focus() should (or even easily can) display the focus in the way that happens when tabbing in to a control.

cnrudd commented:
> This looks great!
> 
> * So we will want to add documented methods signatures (with empty implementations) for all of these on HoistInputModel itself. - Done
> * Also would like to provide mobile implementations -Done
> * interesting question as to whether the imperative focus() should (or even easily can) display the focus in the way that happens when tabbing in to a control.  
- I looked into the focus question.  The focus on the slider, switch and checkbox is prevented by this line:
https://github.com/xh/hoist-react/blob/46ac834704dafb67635041d087dd8455ce941acc/kit/blueprint/index.js#L14
Comment it out, and those components show focus.
I think the solution is to give those components css styles similar to the other inputs that bypass blueprints focus styling overrides.  Let me know if you want me to look into that.

cnrudd commented:
@lbwexler I got the focus outlines working slider handles, checkbox and switch inputs.

cnrudd commented:
I've added focus & blur to Radio Groups & ButtonGroupInput on Desktop.
I've added focus & blur to Checkbox, Switch, and ButtonGroupInput on Mobile.  (Mobile doesn't have a RadioGroup)
2 Issues on Mobile:  

1. There doesn't seem to be a way to to select the content of a select input, so maybe don't implement the select() method.
2. There doesn't seem to be a way to put onFocus, onBlur event handlers on Onsen Buttons, so I had to put the handlers on the buttonGroup hBox.  These handlers are necessary to get the input model's hasFocus property to change accordingly.

cnrudd commented:
@lbwexler As discussed, I've reverted the css/style changes, and moved those to this branch:  https://github.com/xh/hoist-react/tree/focusBlur

lbwexler commented on desktop/cmp/input/ButtonGroupInput.js:
```diff
@@ -49,6 +50,33 @@ ButtonGroupInput.hasLayoutSupport = true;
+    }
+
+    get enabledButtons() {
+        const btns = ReactDOM.findDOMNode(this.domRef.current)
```
Yikes!  This seems like a mistake.  This should be built-in.  Thought domRef was already returning a ref to the dom node1

lbwexler replied:
Let's go over these implementations.  Not sure I fully understand this, but I see that it is complex.

lbwexler commented on desktop/cmp/input/RadioInput.js:
```diff
@@ -95,7 +120,15 @@ const cmp = hoistCmp.factory(
+                },
+                onBlur: (e) => {
+                    model.onBlur(e);
+                    opt.onFocus ? opt.onBlur(e) : null;
```
Are these call backs intended to be a new API on the options (and buttons?).  Also, looks like a copy/paste error here.

cnrudd replied:
Fixed the copy/paste error.
The callbacks are not intended to be "new" on the button elements.  They've always been there, haven't they?
They're just necessary to make sure the button group model is notified that it has focus if one of its buttons is focused, same re blur, but a little more complicated with blur, since blurring one button does not mean the buttongroup is blurred, if another button is focused.

lbwexler commented on desktop/cmp/input/SwitchInput.js:
```diff
@@ -42,6 +44,23 @@ SwitchInput.propTypes = {
 //-----------------------
+class Model extends HoistInputModel {
+
+    inputRef = createObservableRef();
```
Think we need to build this inputRef into the Base class at this point.  Its going to dry up a ton of code.

lbwexler commented:
cool -- sorry if I misunderstood there

On Sat, Dec 12, 2020 at 1:51 PM Colin Rudd <notifications@github.com> wrote:

> *@cnrudd* commented on this pull request.
> ------------------------------
>
> In desktop/cmp/input/RadioInput.js
> <https://github.com/xh/hoist-react/pull/2221#discussion_r541725291>:
>
> > @@ -95,7 +120,15 @@ const cmp = hoistCmp.factory(
>                  disabled: opt.disabled,
>                  label: opt.label,
>                  value: opt.value,
> -                className: 'xh-radio-input-option'
> +                className: 'xh-radio-input-option',
> +                onFocus: (e) => {
> +                    model.onFocus();
> +                    opt.onFocus ? opt.onFocus(e) : null;
> +                },
> +                onBlur: (e) => {
> +                    model.onBlur(e);
> +                    opt.onFocus ? opt.onBlur(e) : null;
>
> Fixed the copy/paste error.
> The callbacks are not intended to be "new" on the button elements. They've
> always been there, haven't they?
> They're just necessary to make sure the button group model is notified
> that it has focus if one of its buttons is focused, same re blur, but a
> little more complicated with blur, since blurring one button does not mean
> the buttongroup is blurred, if another button is focused.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/xh/hoist-react/pull/2221#discussion_r541725291>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARTL2NSNX5DT5JCVEAV553SUO3TJANCNFSM4UOQ7ZBQ>
> .
>

cnrudd commented:
@lbwexler very nice refactor!

amcclain commented on CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@
   `filterFn` prop.
 * New Admin Console Monitor > Memory tab added to view snapshots of JVM memory usage. (Requires
   Hoist Core v8.7 or greater.)
+* Focus, blur, and select methods now available on the model returned by the `ref` prop on most HoistInputs.
```
```suggestion
* Focus, blur, and select methods now available on the model returned by the `ref` prop on most HoistInputs.
* `FormModel` and `FieldModel` also gain new focus-related APIs.
```

amcclain commented on cmp/input/HoistInputModel.js:
```diff
@@ -255,10 +298,10 @@ export class HoistInputModel {
     containsElement(elem) {
-        const thisElem = this.domRef.current;
-        if (thisElem) {
+        const {domElem} = this;
```
Looks like this should be `domEl` ?

amcclain commented on mobile/cmp/input/Select.js:
```diff
@@ -130,6 +130,24 @@ class Model extends HoistInputModel {
+    // maybe not necessary, anyways, on mobile?
+    select() {
+        this.selectText();
+    }
```
I agree that we could leave this clearly un-implemented on mobile

amcclain reviewed (commented):
Hi - looks great, thank you.  I gave it a careful read here on Github - would suggest a CHANGELOG line re. the Form/FIeldModel updates + I think we have one typo - otherwise I am all for merging.
